modules/rubiconAnalyticsAdapter.js

```diff
@@ -0,0 +1,484 @@
+    'dealId',
+    'status',
+    'mediaType',
+    'dimensions', () => utils.pick(bid, [
+      'width',
+      'height'
+    ]),
+    'seatBidId',
+  ]);
+}
+
+let samplingFactor = 1;
+let accountId;
+// List of known rubicon aliases
+// This gets updated on auction init to account for any custom aliases present
+let rubiconAliases = ['rubicon'];
+
+/*
+  Checks the alias registry for any entries of the rubicon bid adapter.
+  adds to the rubiconAliases list if found
+*/
+function setRubiconAliases(aliasRegistry) {
+  Object.keys(aliasRegistry).forEach(function (alias) {
+    if (aliasRegistry[alias] === 'rubicon') {
+      rubiconAliases.push(alias);
+    }
+  });
+}
+
+let baseAdapter = adapter({analyticsType: 'endpoint'});
+let rubiconAdapter = Object.assign({}, baseAdapter, {
+  enableAnalytics(config = {}) {
+    let error = false;
+    samplingFactor = 1;
+
+    if (typeof config.options === 'object') {
+      if (config.options.accountId) {
+        accountId = Number(config.options.accountId);
+      }
+      if (config.options.endpoint) {
+        this.getUrl = () => config.options.endpoint;
+      } else {
+        utils.logError('required endpoint missing from rubicon analytics');
+        error = true;
+      }
+      if (typeof config.options.sampling !== 'undefined') {
+        samplingFactor = 1 / parseFloat(config.options.sampling);
+      }
+      if (typeof config.options.samplingFactor !== 'undefined') {
+        if (typeof config.options.sampling !== 'undefined') {
+          utils.logWarn('Both options.samplingFactor and options.sampling enabled in rubicon analytics, defaulting to samplingFactor');
+        }
+        samplingFactor = parseFloat(config.options.samplingFactor);
+        config.options.sampling = 1 / samplingFactor;
+      }
+    }
+
+    let validSamplingFactors = [1, 10, 20, 40, 100];
+    if (validSamplingFactors.indexOf(samplingFactor) === -1) {
+      error = true;
+      utils.logError('invalid samplingFactor for rubicon analytics: ' + samplingFactor + ', must be one of ' + validSamplingFactors.join(', '));
+    } else if (!accountId) {
+      error = true;
+      utils.logError('required accountId missing for rubicon analytics');
+    }
+
+    if (!error) {
+      baseAdapter.enableAnalytics.call(this, config);
+    }
+  },
+  disableAnalytics() {
+    this.getUrl = baseAdapter.getUrl;
+    accountId = null;
+    baseAdapter.disableAnalytics.apply(this, arguments);
+  },
+  track({eventType, args}) {
+    switch (eventType) {
+      case AUCTION_INIT:
+        // set the rubicon aliases
+        setRubiconAliases(adapterManager.aliasRegistry);
+        let cacheEntry = utils.pick(args, [
+          'timestamp',
+          'timeout'
+        ]);
+        cacheEntry.bids = {};
+        cacheEntry.bidsWon = {};
+        cache.auctions[args.auctionId] = cacheEntry;
+        break;
+      case BID_REQUESTED:
+        cache.auctions[args.auctionId].referrer = args.refererInfo.referer;
```

Comment:
so for a given auction, each bidder gets a `BID_REQUESTED` event.

So we are going to overwrite this a couple times.

I _think_ it will just be the same value every time so probably won't be an issue.

Just wanted to note it.

Also, I am not 100% sure `refererInfo.referer` will always be there, but if the Prebid thing says it will be then we are fine.

Reply:
The mock data in the spec code already had `refererInfo` included in the `AUCTION_INIT` event.

Is that event only fired once? Will that always include this data?